### ISSUE TITLE
Inline hints are too intrusive

### ISSUE THREAD

**Issue by dsyme** [ionide/ionide-vscode-fsharp#1693]
I think the inline type annotations are too intrusive as things stand, it just feels like the highlighted grey boxes are the majority of what I see. I'll keep an eye on it but my gut feeling is 
* The emphasis box around the annotation is visually intrusive and takes attention away from the actual code
* We shouldn't be annotating variables bound in pattern matching by default.
* We should indicate how to toggle the annotations on/off as a hover-tip over the annotations themselves 

e.g. in this code:

![image](https://user-images.githubusercontent.com/7204669/164155104-77540502-1aee-4e1f-9316-6a9bc5df052f.png)

or this code

![image](https://user-images.githubusercontent.com/7204669/164155149-9cb6f364-f747-4a3b-9dfb-b61321451b1e.png)

**Comment by baronfel** [comment]
This has been a common request, but it is documented already. What we could add is mention of the relevant config options directly in the docs about inlay hints, instead of only in the full config options listing.

Separately, there are too many hints reported at the moment, type hints for already typed expressions are unnecessary and will be fixed in the next release.

Finally, vscode itself will be changing the default styling for these hints to a much more subdued default in the next version, which should help immensely.

**Comment by dsyme** [comment]
Adding parameter and union-case names by default looks like too much - a List.map is now hard to read:

![image](https://user-images.githubusercontent.com/7204669/164156174-36b2026b-0e4f-43a1-bf13-c914baca8026.png)

**Comment by dsyme** [comment]
Note I'm really only concerned about the defaults - I don't configure things like this, the defaults have to be right.

**Comment by forki** [comment]
I personally went with type hints on and parameter name hints off. Looks
much quieter.

**Comment by cartermp** [comment]
I expect there to be a few growing pains here. Fundamentally, I think inlay hints are more powerful than tooltips. You get the type definitions for a whole viewport at once, which can really help in many situations, such as a beginner accidentally forgetting to apply all parameters to a function call and having that impact stuff downstream in unexpected ways. That kind of scenario just wasn't dealt with well prior to inlay hints. And since they're just little blocks in the editor, you can interact with them (pop a full tooltip, use it to add a type annotation, etc).

In Visual Studio, they're off by default in C# and need a toggle (alt+f1) which all but kills them as a feature because they're hidden by an obscure key command. But the baby need not be thrown out with the bathwater - I think the right model is actually toggling them on when you need them.

Is there a VSCode story for toggling them on/off?

Separately, there's some questions specific to Ionide about what to keep on or off by default. Inlay hints + line lenses is redundant and we need to pick one and stick with it. Perhaps also true for pipeline hints and codelens type hints as well. What's the best configuration of these things that feels right?

**Comment by Krzysztof-Cieslak** [comment]
> The emphasis box around the annotation is visually intrusive and takes attention away from the actual code

Unfortunately, we don't control how the annotation is displayed - it's handled on the VSCode side

> We shouldn't be annotating variables bound in pattern matching by default.

TBF, I'm not really sure if I agree here, posted screenshots seem fine to me ;-) 

> Is there a VSCode story for toggling them on/off?

There can be toggled on/off with the `View: Toggle Inlay Hints` command. And I do believe VSCode will add a toggling on/off for the current line in the near future.

---

But generally speaking - I agree with what @cartermp said above. There will be growing pains here. It's definitely more intrusive than whatever we had before in F# tooling in VSCode or VS (I do believe Rider had some inlay hints already?). But I do believe it's a good feature (and a direction other tools will move in - TS tooling in Code has been using them for a while, and now it's a public API in Code, so I expect more languages adding support for it). I'd be careful in judging a feature only after a short time of playing with it - LineLenses/CodeLenses were also initially judged as intrusive and after the people got used to them there are now treated as a core feature.

I definitely don't want to turn them off by default - I think we all know that people don't explore options, and never enable opt-in features. I'd rather have them on by default and provide (more) options for the users to configure the experience - TS in Code has waaaaay more options here to control what exactly are hints that should be displayed. We may explore adding more options too, which will also give us a more granular way of controlling default UX (for example "[Adding parameter names by default looks like too much - a List.map is now hard to read" ](https://github.com/ionide/ionide-vscode-fsharp/issues/1693#issuecomment-1103483226) could be an option [and yes, I agree that it potentially could be turned off by default])

> Inlay hints + line lenses is redundant and we need to pick one and stick with it. Perhaps also true for pipeline hints and codelens type hints as well.

This is a good point indeed. Definitely, something we want to explore more.

**Comment by baronfel** [comment]
One potential place where we can clean up duplicated data is to make LineLens and Type InlayHints mutually exclusive. One note about this is that currently InlayHints don't surface the return type of a function, whereas LineLens do. The distinction can be seen here: 

![image](https://user-images.githubusercontent.com/573979/164237842-36687a48-04c6-4109-bace-4300fd41990f.png)

So a potential enhancement to 'truly' replace LineLens with InlayHints might be to consider what an InlayHint for function return types might look like.  Adding another `: int` at the end of the current text wouldn't be correct syntax, so what _should_ we do?

Overall, I think we're on a journey to a) convert our current inlays/hints/decorations to the new-available InlayHint and Decoration Provider APIs, and b) during/after that process to a UX pass on how the whole set of features interacts with each other.

**Comment by dsyme** [comment]
Well, I honestly couldn't read or edit single piece of compiler code with this feature on with its default settings - my productivity hit zero and I switched back to VS. And I'm sure I wasn't about to get used to it. The default settings just feel a fair way off where they need to be. The default should be very minimal type hints - the line lens things were about right. 

The whole point of readable F# code is that the types are non-intrusive, but still rigidly checked. And names like "mapping" for List.map are not even helpful nor form any significant part of the programming model....

Anyway I'd encourage a really proper round of user testing and feedback on this.

**Comment by dsyme** [comment]
Even in the "let add l r =... " example above my eyes gravitate straight to the type annotations, even more than the code. The visuals are much stronger than the line type hints. I find myself barely looking at the actual code at all.

**Comment by dsyme** [comment]
I think the fact that the hints use invalid syntax "let add l : int r : int = ... " (without parents for args) bugs me - if it's an annotation why not use some unicode character instead of ":"? That might help distinguish them, and not make the user read them as code? F# code is different enough without making people learn to read invalid code too. 

Do TS and others inject invalid syntax?

**Comment by dsyme** [comment]
Likewise the List.map example is injecting invalid syntax. Why inject "mapping =" if the "=" is invalid? Won't this surely confuse beginners into thinking they can used named arguments here?

**Comment by cartermp** [comment]
FWIW @dsyme when I used these in my implementation for VS (still in PR), the parameter names were extremely helpful when looking at compiler code and seeing which random `false` value applied to which parameter. One could argue the "true fix" is to fix the code itself, but uhh, that's why tooling can be helpful 🙂

**Comment by dsyme** [comment]
I think the challenge is to make it both opt-in and discoverable

I think sprinkling "Use XYZ to reveal more type information" throughout tooltips and auto-complete would make it fairly discoverable.

> There can be toggled on/off with the `View: Toggle Inlay Hints` command. 

I can't see this command in VS Code - has it got another name?  Or perhaps only in preview?

**Comment by baronfel** [comment]
> I can't see this command in VS Code - has it got another name? Or perhaps only in preview?

It's available in the insiders build of VS Code and so should be dropping in a couple weeks (guessing based on their release history):
![image](https://user-images.githubusercontent.com/573979/164367115-c537d9b9-e27a-42f2-8151-1e9e8849f547.png)  

I think the toggle feature in the Insiders build will be perfect for a lot of folks.  With `editor.inlayHints.toggle` set to "show", hints are shown but suppressed with Ctrl+Alt is held. With that same property set to "hide", the opposite is true. I think it would be very easy to add tooltips to the hints to tell users about these capabilities:

![image](https://user-images.githubusercontent.com/573979/164368213-5c820eaa-6bee-4839-924e-fa7b4a42f177.png)

With a bit more work, we can actually add links to the tooltips that will trigger VSCode commands, and we can use those commands to set these properties on behalf of the user.  With tooltips this long, we'll probably _also_ want to have a setting to tell us it's ok to not show the long tooltip anymore.

**Comment by piaste** [comment]
I think a lot of the intrusiveness actually comes from VSCode's default hint style, which is awful.

For comparison, I reduced the font size and thickness, removed the background color, and dulled the font color, and the result is much more pleasant even with parameter names enabled:

![image](https://user-images.githubusercontent.com/12280683/164406154-48df2b10-3fc2-40de-847d-e46a99391c22.png)

Annoyingly these options can't be set in the UI, but here's the settings.json block:

```
    "editor.inlayHints.fontSize": 11,
    "editor.inlayHints.fontFamily": "Fira Code Light",
    "workbench.colorCustomizations": {
        "editorInlayHint.background": "#FFFFFF00",
        "editorInlayHint.foreground": "#bbbbbb"
    }
```

I don't suppose it would be possible to define an override only for F# files and ship it with Ionide? 

If not, the issue appears to have already been brought up and the default mode should be made somewhat 'compact' soon:

https://github.com/microsoft/vscode/issues/145191

**Comment by forki** [comment]
Unfortunately I now had to disable the feature completely because it delays typing and moves the cursor.

**Comment by baronfel** [comment]
I did some work last night fleshing out tooltips for these in action (In VSCode 1.67.0 and greater there's an additional tooltip to configure the 'toggle' behaviors mentioned above):

![inlay-hint-toggles](https://user-images.githubusercontent.com/573979/164486587-cd246fee-1bf7-4ddf-9416-c52dbe44315c.gif)

@piaste I don't think we can easily scaffold out those settings (they are highly theme-dependent), but we could potentially directly link to docs on how to set inlay hint customizations like that. What do you think?

**Comment by piaste** [comment]
> @piaste I don't think we can easily scaffold out those settings (they are highly theme-dependent), but we could potentially directly link to docs on how to set inlay hint customizations like that. What do you think?

Eh, it's better than nothing, but I don't think it's great UX since those customizations requires manual editing of settings.json. VSCode is a tool for developers, but still.

Changing user settings programmatically from an extension is apparently [possible](https://stackoverflow.com/questions/64628565/how-to-update-users-settings-json-in-vscode-with-extension-packs), and the biggest reduction in intrusiveness comes from setting the background to transparent and making the font size 2pt smaller than the regular text size, neither of which is affected by the theme.

A major issue, however, is that the background colour falls under the 'workbench' key so it cannot be set per-language, and I don't think Ionide should touch that without user permission (maybe add a setting with a warning "this will affect all editors, not just F# ones"?). Reducing the font size, by contrast, can be put under the `[fsharp]` key so I think it would be OK for Ionide to apply a smaller default size, if the user hasn't already manually set it.

**Comment by baronfel** [comment]
Folks on this thread may also be interested in discussion on the FSAC side, which is over at https://github.com/fsharp/FsAutoComplete/issues/924

**Comment by daz10000** [comment]
At risk of adding to the noise - I have to say that the parameter inlays are unusable.  It took me a while to work out why I was having some much trouble reading code (didn't help that I'm temporarily on a different machine with smaller screen) and then I worked out what was doing on.  Every single parameter has a wingman occupying a lot of screen real-estate.  It's hard to tell parameter and type apart (often the same name    `timeout : Timeout  foo  Foo  bar Bar ` - I'm more than happy to hover if I want more context.   Took me 15 mins of searching to find out what these where and similar amount of time to work out how to switch them off..

This might be an unfortunately repetitive example but this is what I'm seeing and the echos are adding nothing and consuming space.

![image](https://user-images.githubusercontent.com/6342645/164949448-3cfb7bc5-f11d-454f-992b-b17f14c13118.png)

**Comment by dsyme** [comment]
Another factor here is that F# explicitly sells itself on succinctness - the subset of people who choose F# are already opting-in to succinct, type-checked code and the loss of visual type information (we usually say "concentrate on domain, not type annotations").  If people wanted fully type-annotated code and non-auto-generalized code then they simply wouldn't be using F#. 

This means the defaults for F# may reasonably be tuned differently to Rust, for example (or any language where there is no HM auto generalization.) I know this cuts both ways - people want succinctness, and the ability to "see more".   

I really want this feature, but I'm pretty sure it needs to be opt-in through a very easy "reveal more information" keystroke (akin to Ctrl-+/- for zoom in and out).  I question whether it should be on by default at all.

Note I have no problem with line `//...` type hints - they are tuned just right and are not intrusive

**Comment by misterspeedy** [comment]
I came back from holiday to find that this feature had appeared, so I had a nice clean before-and-after view of the world.

I really *really* respect the work that has gone into it, but I have to say that I agree with @dsyme that it is more of a distraction than a help when authoring code.  I honestly tried to like it, but ended up wanting to disable it.

In the short term, it should definitely be disabled by default IMO. 

I wonder if longer term there is a case for having two 'views' in VS Code, a bit like Word: a "reading view" and an "editing view".  Rather than having to toggle various such features individually.  Obviously that's a bigger scope thing.

**Comment by baronfel** [comment]
@misterspeedy the toggle command is in the next version of VSCode, 1.67.0. you can see it on the insiders build if you'd like to try it out.

**Comment by dsyme** [comment]
TBH I feel we need to turn this off by default and do an update release - VSCode needs to be a usable pathway for both editing and reading F# code, and my gut feeling is it's just currently not usable (some people may be using it happily, but @misterspeedy and @daz10000 say it's not there).  

Or perhaps this:

1. Turn off by default until the toggle is in non-preview VSCode
2. Remove the "mapping =" insertion of argument names by default
3. Remove any annotation of arguments of curried functions (like [this example](https://github.com/ionide/ionide-vscode-fsharp/issues/1693#issuecomment-1107665402))
4. Somehow advertise how to turn it off

**Comment by baronfel** [comment]
We've done an update to Ionide that adds in the tooltips to let users easily toggle off certain kinds of tooltips, as well as a bunch of changes to FSAC to limit scenarios where parameter tips will apply.  This looks _much_ less cluttered to me now.

We will continue to iterate in this space, but I remain convinced that we want this feature on by default. Now that there is a one-click solution to toggling them off for users that desire that (see [here](https://user-images.githubusercontent.com/573979/164369379-b71f526e-452a-40d4-a8af-bdc78289ac7f.png) for an example), I don't think this is too onerous of a hurdle.

One thing we will likely do very soon is disable the insertion of type hints as @dsyme suggested. Once the LSP API stabilizes for InlayHints, the LSP server itself will be able to supply all necessary text changes to insert an annotation, and it will be able to calculate those on demand instead of guessing ahead of time, and so the overall user experience for those insertions should be much better.

**Comment by baronfel** [comment]
VSCode 1.67.0 has been released, and Ionide 6.0.3 alongside it, which has the nice tooltip to enable the InlayHints toggle when pressing Ctrl+Alt.

**Comment by dsyme** [comment]
@baronfel Thanks! I'll give it a spin

**Comment by dsyme** [comment]
It's a big improvement!

Here's my position: **the landing zone for the majority of users should be to teach them to use Ctrl-Alt**.  That is, with this off by default but Ctrl-Alt known to the user.  This feature in its Ctrl-Alt mode is AMAZING and we should be landing users in the zone where they know how to use this feature.

For example, looking through various files, there are lots of places that feel like a wall of inserted text is shown to me, all of it available via QuickInfo hover tips:

![image](https://user-images.githubusercontent.com/7204669/167903870-3fb1b0d4-80b3-4b9d-bc43-3e764109f592.png)

This wall of text just sucks when it is on by default, and is hopeless when trying to edit code.  On the other hand, the wall of text is **great** when viewed with Ctrl-Alt.  

So for me, the really magical landing zone here for F# code is to teach the user about Ctrl-Alt.  I'll even accept initially turning this on by default just to force the user to learn Ctrl-Alt.  There are very few features I'd say that about but this is one of them.

My recommendation is this: The hints get shown by default **initially** but the intention is 100% to teach the user about Ctrl/Cmd-Alt and then turn them off.

Suggest this:

> "Hey, see these amazing annotations!? They're showing you lots of inferred information about F# code! This is a new feature to help you understand F# code!  You normally access these hints by pressing `Ctrl/Cmd-Alt`.  Got that?
> Click here to hide these hints and start using Ctrl/Cmd-Alt. 
> 
> This will make this dialog go away forever!  If you want to always have these hints on, click here." 

etc.

**Comment by dsyme** [comment]
So my recommendation is this: The hints get shown by default **initially** but the intention is 100% to teach the user about Ctrl/Cmd-Alt and then turn them off.

Suggest this:

> "Hey, see these amazing annotations!? They're showing you lots of inferred information about F# code! This is a new feature to help you understand F# code!  You normally access these hints by pressing `Ctrl/Cmd-Alt`.  Got that?
> Click here to hide these hints and start using Ctrl/Cmd-Alt. 
> 
> This will make this dialog go away forever!  If you want to always have these hints on, click here." 

etc.

**Comment by baronfel** [comment]
yeah, many settings can be, by nesting them in a language-specific node:

```json

"[fsharp]": {
	"editor.inlayHints.enabled": "offUnlessPressed"
}
```

we _should_ be able to do this via the automatic configuration we have setup, it's probably just something wrong with our F# bindings to the VSCode configuration API.

**Comment by baronfel** [comment]
The recommendation is good - instead of relying on pop-up message boxes for things like this, I'd been looking at making use of VSCode's [walkthroughs](https://code.visualstudio.com/api/references/contribution-points#contributes.walkthroughs) feature as a guided demo/introduction for major new features like this. I also want to be able to do a walkthrough for the scenario where we try to find a `dotnet` binary and don't locate one, so we can walk the user through the steps of setting the appropriate config setting so we can even launch.  Want to take a look at those docs and see if you like that solution for inlay hints?

**Comment by dsyme** [comment]
Interesting! Sounds plausible for sure

**Comment by isaacabraham** [comment]
Note that VS Code now supports turning off inlay hints and for you to turn them on "on demand" using a keyboard shortcut.

<img width="344" alt="image" src="https://user-images.githubusercontent.com/1781813/178157310-c467f0a5-04f6-4988-9c88-7ab6e591dfad.png">

Whilst I agree it's quite a lot of information when it's on, I think that having this ability to turn on/off might alleviate the concern.

**Comment by Krzysztof-Cieslak** [comment]
OK, I think with the settings, additional tooltips and the UX improvements done on the VSCode side (`offUnlessPressed`), we're in a pretty good spot. I'd also imagine with the feature being more adapted in general, people will naturally learn about it and its configuration.

Thanks, everyone for the feedback. And please let us know if we need some more changes in this area, but I'm closing the issue for now.

**Comment by boeremak** [comment]
I found this feature today, CTRL + ALT is a godsend. The default is a bit annoying because sometimes the tooltip disappears when you want to hover your mouse over it and you can't click something. Guess why I found this thread. :)

Otherwise: nice work.

**Comment by Beyondo** [comment]
I'd prefer if those hints appear by just moving the cursor over; but them being inline is just so annoying!! Who's the genius who came up with such an idea?? I use the keyword `auto` all the time to shorten code **so I can focus** on the bigger stuff like the algorithm itself and these "inline hints" literally destroy my whole purpose of using `auto`   by making the code look much bigger than it is and they totally distract me... it doesn't stop with `auto` but them appearing literally everywhere. **Make it so they only appear if you hover.**

**Comment by baronfel** [comment]
Hi @Beyondo - if you don't like the feature, have you tried hovering over one of the hints and clicking the link to turn them off? We know this is a feature that not everyone will like in it's most verbose form, which is why we've included a number of customization options for it. And remember you can set these options globally, and even use Settings Sync so that they follow you across editors.

**Comment by cartermp** [comment]
@Beyondo this is also not a place for unproductive rants. Don't post here to insult the hard work of others.

**Comment by Beyondo** [comment]
@cartermp I know I might have let my feelings take a bit over but claiming my comment is unproductive is very not true. I stated clearly why it is redundant and should be removed as you could just hover over the variables you wish to inspect instead of transforming your 3 lines of code into 3 formal paragraphs (you know how long templated C++ types can be). Deeply sorry about my "ranting" though, please forgive me for that and to everyone who has been involved in this feature.

**Comment by dsyme** [comment]
@baronfel @cartermp What's the situation with scripts at the moment?  I don't seem to have a way to turn off inline hints for scripts (so they only activate with Ctrl-Alt)?  

Example: open any script under "docs" in FSharp.Data repo

**Comment by baronfel** [comment]
We don't differentiate in any way between scripts and source files for the hints - it's all controlled by the same flag.

**Comment by dsyme** [comment]
>  We don't differentiate in any way between scripts and source files for the hints - it's all controlled by the same flag, which should be "editor.inlayHints.enabled": "offUnlessPressed".

Hmmm OK, but clicking on the link in the tooltip doesn't seem to add that entry globally?  It tries to add it locally somehow?

**Comment by dsyme** [comment]
That is, I have this in the .vscode/setting.json two directories above the script:

```
{
    "editor.inlayHints.enabled": "offUnlessPressed"
}
```

Maybe because the script is not part of the project Ionide is not paying attention to this setting?

**Comment by dsyme** [comment]
The same seems to apply to any out-of-project source, e.g. if I just open a file like this:

    code tests\FSharp.Data.Core.Tests\NameUtils.fs

Where if I open 

    code .

then the setting is respected.

**Comment by baronfel** [comment]
Ok that makes perfect sense - we could check the current workspace state, and if we're in a 'loose' workspace set the setting globally (at least for the 'fsharp' context).

**Comment by dsyme** [comment]
> Ok that makes perfect sense - we could check the current workspace state, and if we're in a 'loose' workspace set the setting globally (at least for the 'fsharp' context).

Yes that makes sense

**Comment by dsyme** [comment]
@baronfel Should I open an issue for that?  thanks :)

**Comment by dsyme** [comment]
BTW I still think we should just default to the "manually activated" state

**Comment by ioancw** [comment]
> Well, I honestly couldn't read or edit single piece of compiler code with this feature on with its default settings - my productivity hit zero and I switched back to VS. And I'm sure I wasn't about to get used to it. The default settings just feel a fair way off where they need to be. The default should be very minimal type hints - the line lens things were about right.
> 
> The whole point of readable F# code is that the types are non-intrusive, but still rigidly checked. And names like "mapping" for List.map are not even helpful nor form any significant part of the programming model....
> 
> Anyway I'd encourage a really proper round of user testing and feedback on this.

I agree. I've switched to using Rider, as I can't do anything with inlay hints - annoyingly the behaviour around switching them off is inconsistent. Clicking the "Hide these hints" doesn't hide them, and then a few weeks later they disappear.

**Comment by ioancw** [comment]
> @baronfel Should I open an issue for that? thanks :)
> 
> BTW I still think we should just default to the "manually activated Alt/Cmd-Shift" state :)

Did this get implemented? It foxed me for a while, as I just use VSCode to edit .fsx files directly.

**Comment by MangelMaxime** [comment]
@ioancw I think by default the inlay hints are displayed. 

But you can change the `editor.inlayHints.enabled` setting to what you prefer

**Comment by daz10000** [comment]
i’ve gotten used to just switching this off every time I hit a new installation. I did. run into a new problem recently on a Mac where the click here to disable button wasn’t clickable so I had to find the setting manually and disable.  It’s not completely obvious which visual feature maps onto which hint but I eventually got rid of them.  I’m sympathetic to the new user experience - it probably would have helped me while learning but the amount of visual overload in the modern IDE is getting crazy.  with copilot installed you have suggested completions,  fsautocomplete is also offering completions , hints are annotation things as well, as you hover you see things. It’s getting harder to see the actual code :)
